sw/device/lib/base/memory.c

```diff
@@ -76,24 +212,85 @@ int OT_PREFIX_IF_NOT_RV32(memrcmp)(const void *lhs, const void *rhs,
 
 OT_WEAK
 void *OT_PREFIX_IF_NOT_RV32(memchr)(const void *ptr, int value, size_t len) {
-  uint8_t *ptr8 = (uint8_t *)ptr;
-  uint8_t value8 = (uint8_t)value;
-  for (size_t i = 0; i < len; ++i) {
-    if (ptr8[i] == value8) {
-      return ptr8 + i;
+  const unsigned char *ptr8 = (const unsigned char *)ptr;
+  const uint8_t value_byte = (uint8_t)value;
+
+  size_t leading_bytes_end, aligned_words_end;
+  compute_alignment((uintptr_t)ptr, (uintptr_t)NULL, len, &leading_bytes_end,
+                    &aligned_words_end);
+  size_t index = 0;
+  for (; index < leading_bytes_end; ++index) {
+    if (ptr8[index] == value_byte) {
+      return (void *)&ptr8[index];
+    }
+  }
+  const uint32_t value_repeated = repeat_byte_to_u32(value_byte);
+  for (; index < aligned_words_end; index += sizeof(uint32_t)) {
+    uint32_t word = read_32(&ptr8[index]);
+    uint32_t bits_eq = ~(word ^ value_repeated);
+    static_assert(__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__,
+                  "memchr assumes that the system is little endian.");
+    if ((bits_eq & 0xff) == 0xff) {
+      return (void *)&ptr8[index];
+    }
+    if ((bits_eq & 0xff00) == 0xff00) {
+      return (void *)&ptr8[index + 1];
+    }
+    if ((bits_eq & 0xff0000) == 0xff0000) {
+      return (void *)&ptr8[index + 2];
+    }
+    if ((bits_eq & 0xff000000) == 0xff000000) {
```

Comment:
nit: please consider the following since it would be slightly easier to review than counting zeros.

`(bits_eq >> 24) * UINT8_MAX == UINT8_MAX`

applies to previous 3 if statements too.

Reply:
Hmm, I don't follow that arithmetic. Suppose `bits_eq` is 0xff because the last 8 bits of `word` matched `value32`. In that case, `bits_eq * UINT8_MAX` is 0xfe01, which would be truncated as 0x01.

But I see your point about having to count zeroes. I'll rewrite it to avoid that cognitive load.

Reply:
Done